File: erlang/pe014.erl
-module(pe014).
-export([start/0]).

max() -> 1000000.

start() ->
  pe014(0, 0, 1, max()).
  
pe014(Maxv, Maxi, I, M) when I < M ->
  Next = pe014(0, I),
  if 
    Next > Maxv ->
      pe014(Next, I, I + 1, M);
    true ->
      pe014(Maxv, Maxi, I + 1, M)
  end;
      
pe014(_Maxv, Maxi, _I, _M) ->
  io:format("~w~n", [Maxi]).
    
pe014(C, N) when N =:= 1 ->
  C;

pe014(C, N) when N rem 2 =:= 0 ->
  pe014(C + 1, N div 2);
    
pe014(C, N) ->
  pe014(C + 1, 3 * N + 1).

File: ocaml/pe014.ml
open Int64

let million = 1000000L ;;

let even x = rem x 2L = zero ;;

let rec collatz c n =
  if n = one then
    c
  else 
    if even n then
      collatz (add c one) (div n 2L)
    else
      collatz (add c one) (add (mul 3L n) one) ;;
     
let rec pe014 maxv maxi i m =
  if i < m then
    let next = (collatz zero i) in
      if next > maxv then
        pe014 next i (add i one) m
      else
        pe014 maxv maxi (add i one) m
  else
    Printf.printf "%Ld\n" maxi ;;

let main () =
  pe014 zero zero one million ;;

let () = main ()
